### DISCUSSION TITLE
Client types - JSON format, configuration examples

### DISCUSSION THREAD

**Discussion by mposolda** [keycloak/keycloak#9066] [Ideas]
This is follow-up of the discussion initial [client types discussion](https://github.com/keycloak/keycloak/discussions/8497) with some more concrete examples of JSON configuration and examples of _global client types_, which will be defined in Keycloak by default.

### Global and realm client types

The configuration of client types would be available in JSON file. Similarly like for client policies or user profile, there are 2 types of client types:

- **Global client types**: predefined client types inside Keycloak. Those are configured in the JSON file inside keycloak-services JAR artifact in the Keycloak distribution. Due the fact they are saved inside the distribution, they don't need to be saved in Keycloak DB, which means they can be easily migrated between Keycloak versions (Changing something in the JSON will automatically mean that new Keycloak version will see this change). However Keycloak administrators cannot change these global client types

- **Realm client types**: configured by administrator for the particular realm. JSON is internally saved as realm attribute (Same like already for client policies)

### Example JSON configuration

The configuration of single client type can look like this:

    {
      "name": "service-account",
      "provider": "json",
      "config": {
        "read-only-attributes": {
          "protocol": "openid-connect",
          "publicClient": false,
          "bearerOnly": false,
          "standardFlowEnabled": false,
          "implicitFlowEnabled": false,
          "directAccessGrantsEnabled": false,
          "serviceAccountsEnabled": true,
          "fullScopeAllowed": false
        },
        "writable-attributes": {
          "protocolMappers": [
            {
              "name" : "Client IP Address",
              "protocol" : "openid-connect",
              "protocolMapper" : "oidc-usersessionmodel-note-mapper",
              "consentRequired" : false,
              "config" : {
                "user.session.note" : "clientAddress",
                "id.token.claim" : "true",
                "access.token.claim" : "true",
                "claim.name" : "clientAddress",
                "jsonType.label" : "String"
              }
            }
          ],
          "defaultClientScopes": [ "roles" ],
          "optionalClientScopes": []
        }
      }
    }

**Name**: name of client type

**Provider**: This will be usually "json" . The Keycloak will provide one built-in java provider "json", which itself will read the configuration of client type from JSON format. Administrators can deploy their own implementation of client type and implement whole "client type logic" in Java for example. In this case, the "config" field can be completely ommitted for such providers or can be in different format

**Config**: There are two parts
- **read-only attributes**: Client of this type will have attributes with those hard-coded values and they cannot be changed. For example clients of type "service-account" have protocol hardcoded to "openid-connect", they always have "service accounts" switch enabled. This enforces behaviour of client in some aspects, which makes sense for the context of this client type

- **writable-attributes**: Client attributes with their default value. Administrator can change/update them.

### Client attributes

The "writable-attributes" and "read-only-attributes" refer to:
- Java properties of the `ClientRepresentation` class (EG. `standardFlowEnabled`, `serviceAccountEnabled`). This includes non-trivial properties like `protocolMappers`, which itself are provided in JSON format

- Attributes of client referred with `clientRepresentation.setAttribute` (EG. attribute like `jwks.url` or `pkce.code.challenge.method`)

- Some properties of client, which are not referred by any clientRepresentation properties or attributes. For example "service account roles" or "scopes" (Scope role mappings in the client "Scope" tab of the client in the admin console). We will need to document what are supported names of these special properties.

Attributes, which are not explicitly mentioned in the "default-attributes" or "read-only-attributes" are considered to be unsupported for the particular client type

### Global client types

Idea is, that client type will be mandatory property of the client. Always when new client is created, it will have "type" assigned. The built-in types for the client might be:

- `openid-connect` - The client with protocol `openid-connect` and with no further restrictions. When we introduce that "client type" is mandatory, the existing clients of openid-connect protocol will be migrated to this client type.

- `saml` - The client with protocol "saml". Migration same like for `openid-connect` above

- `service-account` - A bit more opinionated `openid-connect` client with enabled `service-account` and disabled other flows related to browser etc. This client type is more opinionated as lots of properties related to typical `openid-connect` client may be hidden from this client

- `openid-connect-prod` - The same like `openid-connect` but with defaults set to be more secure by default. For example disabled `Full Scope Allowed` and `Direct Access Grant` login (OAuth2 resource owner password credentials grant)

### Admin console - client types

Client types can be edited in admin console in two formats:

- **JSON editor**: Administrator can directly push JSON configuration of his client type. Useful when copying client types from environment to another or when using some existing client type as "template" (JSON can be copied from existing client type and then just few requested things changed etc)

- **Form view**: More user-friendly editor. Client type configured by administrator is internally translated to the JSON (Similar approach to Openshift (admin console VS directly push YAML) and/or to Keycloak client policies)

### Admin console - client view

Client view will be able to read metadata from client type. Based on that, some fields/tabs, which are not supported by particular client type can be hidden for the particular client type. Other fileds/tabs are read-only.

Overally, client view can be more opinionated about what makes sense for the particular client type.

### Remove/Simplify existing Keycloak concepts?

Client types may allow us to remove/deprecate some existing Keycloak concepts and replace them by client type. Some examples:

- **Realm default client scopes**: Currently in the tab "Client Scopes" -> "Default Client Scopes" (URL like http://localhost:8081/auth/admin/master/console/#/realms/master/default-client-scopes ) administrator can configure what are "Default default client scopes" and "Default optional client scopes", which are added to newly created clients. Client types may allow to remove this concept as it can be directly defined on the client type, which client scopes should be added by default.
  
  So for example `openid-connect` client type will define `email` and `profile` as `Default client scopes` for this client type and    `address` and `phone` as Optional client scopes.
  
  This allows us that client types like `service-account` can have different client scopes added to them as for example `email`, `phone`, `profile` or `address` don't have much sense for service accounts.

- **Hardcoded protocol mappers for service-accounts**: Currently when service-accounts are enabled for the client, some protocol mappers are added to such client. This can be removed and protocol mappers like `Client IP Address` can be added to `service-account` client type instead. Or possibly moved to some client scope, which can be added as "Default client scope" of service account client (as we move toward removing of "Protocol mappers" and "Role scope mappings" from client itself and have it on client scope instead)

### Versioning

Client types will support versioning. This can be useful especially for backwards compatibility.

For example we have client switch `exclude-session-state-from-auth-response`, which is supposed to be disabled by default. Only for backwards compatibility, we may need to enable this switch in the legacy clients. Hence the client type `oidc` may have this switch disabled in latest version 2. In the older version 1, the switch is still enabled.

Note that most of the client fields definitions are defined in the latest version and they are not repeated for every version. Including the attribute definitions in latest version has advantage that administrator can easily see what are applied fields in the latest version as 99% of cases will use latest version. Hence administrator doesn't need to "decode" by looking at the definitions of all the old versions. The older version 1 contains definitions just for the fields, which are different than in the latest version.

Here is the example how can the definition of both versions look like (For attribute definitions, example uses variant 5 of [this comment](https://github.com/keycloak/keycloak/discussions/9066#discussioncomment-1815094), which is my current preference):
```
{
  "type": "oidc",
  "version": "2",
  "latest-version": true,  
  "attributes": {
    "protocol": {
      "value": "oidc"
    },
    "standardFlowEnabled": {
      "default-value": true
    },
    "baseUrl": {},
    // ... All OIDC attributes defined here
    
    // Compatibility attribute
    "exclude-session-state-from-auth-response": {
      "value": false
    }
  }
},
{
  "type": "oidc",
  "version": "1",
  "latest-version": false,
  "attributes": {
    "exclude-session-state-from-auth-response": {
      "value": true
    }
  }
}
```

When the client is being created, administrator can choose any of types `oidc`, `oidc:1`, `oidc:2`. Type `oidc` assumes to refer to the latest version. If administrator explicitly choose `oidc:1`, client will refer to version 1.

### Inheritance

We can support inheritance of client types.

The client type can have `parent` field. Parent field can be for example just `oidc`, which refers to latest version. Or `oidc:1`, which would refer to `oidc` type of version 1.

There will be cases when child client type won't support all the fields, which are supported by parent type. For example assuming that type `service-account` will inherit from type `oidc`. However client type `oidc` supports lots of attributes and many of them applicable just for "browser" use-cases (EG. `consentRequired` field) . So `service-account` will need some way how to exclude some attributes, which are defined on parent. The possibilities are:
- Add the field `excluded-attributes` to the `service-account` client type and list the excluded attributes. This has disadvantage that  most of child client types will need to be changed when new field is added to the parent.
- Add field `inherited-attribute-definitions`, which will list attributes, which are inherited from the parent. This is probably better option than the `excluded-attributes` and I am using it in the example below
 
The child type will inherit all the attribute definitions from parent. It will need to re-define just some attributes, which are different than parent. Added an example below for `service-account` (example uses variant 5 of [this comment](https://github.com/keycloak/keycloak/discussions/9066#discussioncomment-1815094), which is my current preference).

```
{
  "name": "service-account",
  "version": 2,  
  "parent": "oidc",
  "inherited-attribute-definitions": [ "protocol", "baseUrl", "..."],
  "attributes": {
    "standardFlowEnabled": {      
      "value": false
    },
    "serviceAccountEnabled": {      
      "value": true
    },
    "useRefreshToken": {
      "default-value": false
    },
    "defaultClientScopes": {
      "default-value": [ "roles" ]
    },
    "optionalClientScopes": {
      "default-value": []
    }
  }
}
```

**Comment by stianst** [comment]
I wonder if there should actually be two JSON representations here:

* Declarative JSON type: used to create/define a client type in JSON
* Client type metadata: used by the admin console when for example editing a client

That may be better than try to mix the two into a single representation. It would also remove the somewhat confusing "provider" field when creating a new client type using JSON.

**Reply by mposolda** [reply]
I personally like the same approach like used for client policies. So administrator can configure things in the "form" with the nice UI in the user-friendly way, but things are internally always represented as JSON configuration.

I see the single JSON format used as easier to understand and implement than two different formats. Another advantage is, that for example administrator can use global/builtin client types as "template" and just copy/paste their JSON configuration and then change "manually" (with admin UI) just few things, which he want to update.

For example Keycloak may have global `service-account` as one of the default client types. When administrator wants to do some small change of it (EG. add one new client scope), he can just copy the JSON of the existing global configuration, then add the client scope via UI and save this new client type.

**Reply by stianst** [reply]
I'm not quite sure. I can imagine there is things that you need to specify when creating a JSON declarative type that differs from the metadata needed by the UI. For one it makes no sense to have a `provider` field when creating a JSON declarative type, as that is implied, and there is no other options, so pointless to confuse people with this field.

**Reply by mposolda** [reply]
The `provider` is the only field, which I can think of. For all the other fields, like individual fields of the client, you will always have a chance to configure them in both UI and JSON and they should displayed in both places.

Going one step back, I am not convinced if we need `provider` field for client type. I mean we can likely stick just with the `json` provider and ommit this `provider` field from both UI and JSON? I think that defining client type in JSON would always work fine. In other words, I am not convinced if we ever need the implementation of client type hardcoded in Java.

When comparing client type definition in JSON comparing to Java, I see the advantages of JSON approach like:

- With JSON, administrator clearly see what client fields would be supported for particular client type and what are their default values. With Java, it is all hidden behind black-box, which is not very useful for typical administrator IMO.

- When we have built-in types like `service-account` defined in JSON, it allows administrator to use them as "template" and copy the JSON of existing `service-account` type and change just one or two fields, which administrators want differently. That is one thing, which is easily doable with client policies and works nicely IMO. With implementation of client type hardcoded in Java, I don't see "template" approach easily doable. 
The "inheritance" support may be an alternative for the "template" approach as admin can inherit from built-in `service-account` type and re-define just the 2 fields, which he wants. But for this guy, it will be still useful to see the original field definitions in JSON so he see what he really inherits from and what are the fields, which he need to re-define.

I don't see any advantage of the black-box approach with the client type definitions in Java TBH. Is there any?

**Comment by stianst** [comment]
Should we consider versioning of client types? "my client" is a "SPA v1", while "my other client" is a "SPA v2"?

**Reply by mposolda** [reply]
Maybe. I am not sure if versioning should be 1st-class citizen or whether it should be managed by administrator himself?

For example administrator can easily add new client type named `My SPA V2` , which will be based on existing type `My SPA` . With JSON config of client types, it is very easy to use existing client type as "template" and just change things, which should be different in that new client type version  (see above https://github.com/keycloak/keycloak/discussions/9066#discussioncomment-1815065 ).

Then administrator can assign `My SPA V2` to clients, which he wants to use the new client type version and keep `My SPA` on the old clients.

**Reply by stianst** [reply]
By making versions a first class citizen what it allows you to do though is to say "any version of SPA", or explicitly "SPA v1". That allows either manually updating to changes in built-in types, or automatically updating.

**Reply by stianst** [reply]
Was thinking about something that client type versioning could be very useful for, which is when we need to break the protocol API.

Right now we have a bunch of compatibility switches available for a client. Now instead of having these exposed on clients directly we could make them an internal thing on client types.

So, if you say lock a client type to "oidc:v1" then you are guaranteed backwards compatibility until that version is removed at some point. Then you can upgrade the client and update to "oidc:v2 at your own leisure. Alternatively, you can just use "oidc" and make sure you read the migration guide that should say there's a new version of "oidc" and list the relevant changes, then you can either fix the client at the same time as the upgrade, or perhaps as a temporary workaround fix it to "oidc:v1".

This might suggest that we may want to have some inheritance of client types? Or, at the least that all custom client types should be children of a root protocol client type like "oidc" and "saml".

**Reply by mposolda** [reply]
Yes, I think that both versioning and inheritance support will be nice when thinking more about this :-) Especially inheritance is needed and will be really useful to avoid duplicated field definitions etc.  I've added new sections **Versioning** and **Inheritance** with some details.

Inheritance is tricky in a way, that child may not support all the fields by parent. Like `service-account` client type won't support all the fields by `oidc` client type. It seems the option is that child will either need to define what are "excluded" attributes, or list those attributes, which it wants to "inherit" . I've added this to the "Inheritance" section above.

**Comment by stianst** [comment]
Not sure I like the terms read-only and writable, but not sure I can think of a better alternative. One question if I want to allow an attribute to be writable, but I don't want to set a default value, what do I do?

**Reply by mposolda** [reply]
Few possibilities for this:

1) Set null:
```
"baseUrl: null"
```

2) Set empty value:
```
"baseUrl: ""
```

3) Wrap the default value inside separate element. So in the case I don't want default-value I would use this:
```
"baseUrl": {}
```
in case I want default value, I would use this:
```
"baseUrl": {
    "default-value": "https://foo"
}
```


4) Variant of 3, which assumes that elements `writable-attributes` and `read-only-attributes` won't be used. Instead the attributes will be enumerated directly under `attributes` and each attribute will contain whether it is read-only (false by default) and default-value (can be ommitted in case default-value won't be used). So something like:
```
"attributes": {
  "standardFlowEnabled": {
    "read-only": true,
    "default-value": true
  },
  "defaultClientScopes": {
    "read-only": false,
    "default-value": [ "profile", "email" ] 
  },
  "baseUrl": {
    "read-only": false
  }
}
```

In this example, the baseUrl is supported and won't have default-value.

5) Variant of 4, which ommits `read-only` field and instead use `value` VS `default-value` . The `value` means that value is hardcoded and hence read-only. The `default-value` means that field is not read-only and client can be changed. Same definition like for 4 would look like this:
```
"attributes": {
  "standardFlowEnabled": {    
    "value": true
  },
  "defaultClientScopes": {    
    "default-value": [ "profile", "email" ] 
  },
  "baseUrl": {}
}
```

My vote is to go with 5.

**Comment by stianst** [comment]
Some use-cases I'm not sure is covered by the current proposal:

* Don't allow defining any protocol mappers
* Don't allow defining any client roles
* Don't allow defining any scopes
 Don't allow defining any client scope
* Only allow a specific list of protocol mappers
* Only allow a specific list of client roles
* Only allow a specific list of scopes
* Only allow a specific list of client scopes

**Reply by mposolda** [reply]
In shortcut, everything you mentioned should be supported.

More details below:

> Some use-cases I'm not sure is covered by the current proposal:
> 
>     * Don't allow defining any protocol mappers

I count with the fact that "protocolMappers" itself is a property of client type. So if it is not defined as an attribute, it means that `protocolMappers` are unsupported property.

Alternative is also to define protocolMappers as read-only attribute and hardcode it to empty list. This effectively means hardcoded empty list of protocolMappers. So we either use something like this in `read-only-attributes` :
```
"protocolMappers": []
```
or use something like this in case we go with proposal 4 from above https://github.com/keycloak/keycloak/discussions/9066#discussioncomment-1815094
```
"protocolMappers": {
    "read-only": true,
    "default-value": []
}
```

> 
>     * Don't allow defining any client roles

Same like for protocol mappers above.

There is one implementation detail, that client roles are not defined as property on the `ClientRepresentation` .  But I assume that client types will support client roles even if client roles are saved in Keycloak outside of client representation / model. We can define property like `client-roles` to handle client roles. I consider it as "special property" similarly like "role scopes" or "Service account roles" (See the proposal in section "Client Attributes" in 3rd bullet).

> 
>     * Don't allow defining any scopes

Same as for client roles. We can have property like `role-scopes` to cover role scopes (Stuff in the tab "scope" of the client)

>       Don't allow defining any client scope

Same as above. ClientRepresentation have proeprties `defaultClientScopes` and `optionalClientScopes`

> 
>     * Only allow a specific list of protocol mappers

Protocol mappers will be defined as read-only property with hardcoded list of protocol mappers. I have this in the prototype for hardcoded read-only list of protocol mappers defined for `service-account` client type: https://github.com/mposolda/keycloak/blob/client-types/services/src/main/resources/keycloak-default-client-types.json#L64-L94

> 
>     * Only allow a specific list of client roles
> 
>     * Only allow a specific list of scopes
> 
>     * Only allow a specific list of client scopes

Same as above points.

**Comment by hmlnarik** [comment]
I wonder if this represents constraints that should be defined on storage level. Is there KeycloakSession used for anything, or is it only list of allowed attributes / fields that can be defined for a particular client type?
It very much resembles the work being done on tree store, specifically the access to the [fields and attributes could be filtered](https://github.com/keycloak/keycloak/blob/main/model/map/src/main/java/org/keycloak/models/map/common/delegate/EntityFieldDelegate.java) by field name and / or attribute name, and could leverage description of the attributes (could leverage what's done for fields, [names of the fields are generated automatically](https://gist.github.com/hmlnarik/01d2459100b1a7a410b6a4b54ed51f74#file-mapcliententityfield-java-L77))

**Reply by mposolda** [reply]
I am not 100% sure about constraints on the storage level, but maybe I don't understand exactly what you mean...

Current assumption is that client-type is the entity defined by administrator. For example administrator can enforce that field `standard flow enabled` of client of client-type `service-account` won't be editable by client administrators. Which should be reflected also in the admin console UI and field should be read-only here. Maybe the idea behind client types is better explained in the original Stian's proposal of client types: https://github.com/keycloak/keycloak/discussions/8497 .

So this enforcement is not in the storage layer as various client types can have field `standard-flow-enabled` editable or not editable.

I currently have this in my current prototype https://github.com/mposolda/keycloak/blob/client-types/services/src/main/java/org/keycloak/services/clienttype/client/TypeAwareClientModelDelegate.java . It is based on the augmentation approach mentioned by @stianst  in the proposal https://github.com/keycloak/keycloak/discussions/8497 . It supports only single field `standardFlowEnabled`, but idea is to have something like this for other client representation fields.

In the current prototype, I integrate with the old store by wrapping `ClientAdapter` instances returned from JPA / Infinispan stores with this "client delegate" based on the "type" set on the client itself.

Not sure if there is some overlap with the new store regarding this? I assume that client types should work with both old and new store. So not 100% sure if I can directly use classes, which you refer?

**Comment by hmlnarik** [comment]
Also I wonder whether we could leverage YAML format for the client types. Client types often have some reasoning which would ideally be captured in comments. There are no means to write comments in JSON, contrary to YAML.

**Reply by mposolda** [reply]
So far we use JSON everywhere in Keycloak and it is also used for client policies and userProfile definitions. JSON is also used by OIDC specification.

On the other hand, maybe YAML is better format for defining things like this? I don't have any issues with introducing YAML and define things in YAML format instead of JSON. This will likely mean that I will need to introduce some "infrastructure" for support parsing/serializing YAML into Keycloak codebase, as we don't yet have anything AFAIK?

@stianst WDYT? Does it worth to introduce YAML support in Keycloak and define client types in YAML instead of JSON?

Side-note: If we decide to stick to JSON, there is possibility to add `description` attribute to client type. However I understand that this won't fully replace flexiblity provided by comments.

**Comment by alechenninger** [comment]
@stianst @mposolda would you consider starting with design for an SPI interface for this? This would enable very simple implementations of types (e.g. hard coded extensions) before a more mature feature is fully developed.

**Comment by patrickjennings** [comment]
One thing that warrants discussion is how this feature will change how the configuration of clients is resolved.

The source of truth for entities in Keycloak has historically been provided through the connected datastores. With how we have this feature defined, that would not be the case. A clients config would be driven through what is stored in the DB/cache as well as through any client type configurations associated.

This has implications on how we obtain clients and client related data across Keycloak. For instance, named queries are used to aggregate client information from the DB, and functions like client attributes search are driven through DB queries. With the augmentation of clients in the app, it's not clear how this would work going forward.

We could try to keep things consistent with the db. Yet if we allow anyone to extend and implement their own client type provider, it might not be possible to do this since the resolution of the client properties could be dynamic. Or even the fact that updating a client type could be very intensive as we would have to update each associated client in the db to make it consistent.

What do you all think about this? @mposolda @stianst @alechenninger @jsorah @vickeybrown

**Reply by mposolda** [reply]
That is interesting concern.

If I understand correctly, if there is attribute `foo` with value `bar` defined on the client type `my-client-type`, then all clients of this client type will inherit this attribute (but they may override it etc). So then there is attempt to search all clients with attribute `foo=bar`, then it should search all clients, but also client types and hence return all clients of the particular client type `my-client-type` (unless they override the value of property `foo` with some different value). Is it the concern you have?

I see that especially this can be problematic with default client types as you pointed. It is possible that `ClientTypeSPI` itself may need some searching capabilities? This can allow the author of dynamic client type provider to have a way to implement searching by himself and return the appropriate client types.

But fortunately, it seems there are not so much places, which are affected by this? I see:
- `ClientLookupProvider.searchClientsByAttributes`
- `ClientLookupProvider.searchClientsByAuthenticationFlowBindingOverrides`
- `ClientProvider.getAlwaysDisplayInConsoleClientsStream`
- `ClientProvider.getAllRedirectUrisOfEnabledClients` (but this one is deprecated).

It is possible we may want something like default `ClientProvider` (something like `ClientTypeAwareClientProvider`?), which will be returned by `KeycloakSession.clients()` . It will delegate to the currently returned `ClientProvider`, but will add the searching capabilities for searching also client types and merging search results from clientTypes and "plain clients" ?

**Comment by jsorah** [comment]
Regarding Global client types

I think at least three of the proposed options make sense and are clear in their implementation:
- openid-connect
- saml
- service-account

I am wondering if we start with the above three and then later create a specific `openid-connect-prod` (or equivalent, profile).

For that profile I could certainly see:
- no implicit flow
- no ROPC flow
- no full scope

But given the amount of variance in defining attributes/properties on a client (SAML or OIDC) and folks having different tolerances for what is "hardened" in their environment - I am wondering if it maybe it should be left up to an administrator to define their idea of a `production`  or hardened type?

Another question I have is how much of this hardening could also be delegated to an appropriate client profile that is enabled based on the client type?

**Reply by mposolda** [reply]
+1 for `openid-connect` , `saml` and `service-account` client types.

I am not 100% sure about `openid-connect-prod`. As you pointed, people can have different requirements and also there is some intersection with client policies. Maybe we can start with those 3 default client types and as a follow-up, see if we need some more?

I suppose we want to have ability for administrators to have a way to add his own "default client types" configuration, which will be available for every new realm? So some way to override `default-client-types.json` file with the custom configuration. But I am not sure about the priority for this. Probably something to be added at the later stage when client-types feature is mostly ready?

**Comment by mposolda** [comment]
Had a discussion with @stianst today. Few suggestions, which we discussed:

- We may not allow realm administrators to create client types. Client types would be always deployed to the server and it would be possible to use only deployed client types when creating client. Realm administrator may have a way to disable some client types for the realm, but not option of adding new client types. This will also limit the count of client types and hence will simplify searching use-cases (the "searching concern" above mentioned by @patrickjennings)

- Client type may prescribe hardcoded value of some options (so EG. client type `service account` will specify that option `Service Accounts Enabled` is always `ON` for the clients of this client type). But client types may not handle "default values" for clients. In other words, when client type specifies the value for some attribute (like `ServiceAccountsEnabled=ON`), the client won't have option to update this value. This will again simplify searching (as otherwise, we would need to handle the case when client type specified default value `foo=bar`, but client updated default value, then search results for "search client by attribute `foo=bar`" would need to filter the clients, which updated the default value).

- Ideas for default client types might be those (to be clarified later as those don't need to be perfect and are subject to change based on the feedback from community etc). Top types are `oidc` and `saml`. The `oidc` can then have those "child" client types (as client types should ideally support inheritance):

  - Server-side web application
  - Single-page application
  - Single-page application with dedicated REST API
  - Service account client
  - Future thing where the SPA invokes external REST API through a dedicated backend-for frontend, which also results in a confidential client and no need to consider CORS
  - Native application
  - REST API
  
- Admin console may hide the options, which are not applicable for particular client type. Also it can hide all the options specified by client type (so EG. client of type `service account` won't need to show the option `Service Accounts Enabled` as it is implicitly enabled that `Service Accounts` is ON for this client type).

Somewhere at the top of the screen, client can have a link to his client type, which will allow administrator to click on the client type and seeing some screen with the "visualization" of the client type (So showing the values of some options etc). We will probably need to discuss this with the UI team and the "design team". It seems the Client Type SPI will need to have a way to return the "representation", which will allow to specify how the representation of this client type should look in the UI (but this is probably low priority)

Client can have options to even hide some sections entirely or hide some tabs (EG. client `single-page application` can hide `Service account roles` tab or `credentials` tab as this client always has client authentication disabled). But this can become complex as lots of those conditions for hiding some tabs are already specified (EG. tab `Credentials` is already shown just for the confidential clients, or tab `Service Account roles` is shown based on the `Service Accounts Enabled` switch etc). So I suppose this is also lower priority (maybe not needed at all?)

**Reply by jsorah** [reply]
Thanks for the update - still processing the impact of some of it, but a few things that I thought of immediately while reading this that I wanted to immediately ask about:

> We may not allow realm administrators to create client types. Client types would be always deployed to the server and it would be possible to use only deployed client types when creating client. 

I think I understand, but just to make sure - any custom client types would be created in some definition deployed via an extension and we wouldn't need to expose an API for dynamic management of the types?  Do you anticipate if this would mean they would always be globally defined (for all realms), or there would be some way to define realm specific ones even in this "extension" only mode?

> Ideas for default client types

Does `bearer-only` make sense as another type? I know some things changed where this went from being something defined with a specific auth setting `bearer-only` to now "turn off all the auth flows for an  oidc client"?

**Reply by patrickjennings** [reply]
Thanks for clarifying.

Question I have is whether we would support augmentation of the `ClientModel` through client types or only the `ClientRepresentation` and augmentation/validation of it during client creations and updates?

There is the notion of the `TypeAwareClientModelDelegate` class which applies client type configuration over an instance of a `ClientModel` returned by Providers. Which I think the idea with this was that an admin would be able to apply changes to client type configurations, and all of the associated clients would automatically see those changes applied. For instance, adding an attribute or protocol mapper would automatically augment all of the clients associated with those changes. Although like I was referring to before, this gets tricky because the source of truth is no longer solely in the database, which the app had previously assumed was true.

If we do not support this and only support the validation / augmentation of client types during create and updates of a client, this simplifies things. With the caveat that we would lose the ability to manage client configurations at scale through the client types and changing an established client type configuration would cause any associated client to lose consistency with it.

**Comment by mposolda** [comment]
> Thanks for the update - still processing the impact of some of it, but a few things that I thought of immediately while reading this that I wanted to immediately ask about:
> 
> > We may not allow realm administrators to create client types. Client types would be always deployed to the server and it would be possible to use only deployed client types when creating client.
> 
> I think I understand, but just to make sure - any custom client types would be created in some definition deployed via an extension and we wouldn't need to expose an API for dynamic management of the types? Do you anticipate if this would mean they would always be globally defined (for all realms), or there would be some way to define realm specific ones even in this "extension" only mode?

We discussed about client types to be deployed as extensions (providers) or as built-in client types. Then those deployed client types would be available to all realms. So far, we discussed the "realm admin" can possibly disable some client types, so the "client administrators" of particular realm won't be able to use all the available deployed types. But realm admin cannot add new client types.

So for instance:
- There will be built-in client types `a`,`b` and deployed client types `c`, `d`, `e` . 
- Realm administrator of realm `foo` does not want clients of his realm to use type `e`. So he would disable it. So clients of his realm have a way to use types like `a`, `b`, `c`, `d`. For the sake of simplicity, the realm admin won't have a way to add his own client type `f` specific to this realm.

WDYT? Would it be ok to you or do you see any issues with this approach? 

> 
> > Ideas for default client types
> 
> Does `bearer-only` make sense as another type? I know some things changed where this went from being something defined with a specific auth setting `bearer-only` to now "turn off all the auth flows for an oidc client"?

Yes, I think "REST API" is probably similar (or same thing) as bearer-only? We can clarify what naming is better, but seems to be the same thing. Basically the client `foo`, which represents the admin REST API, which can be accessed by the tokens from other clients. But this client `foo` himself cannot authenticate against Keycloak and obtain his own tokens.
